packages/rpc-graphql/src/resolvers/resolve-info/account.ts

```diff
@@ -0,0 +1,95 @@
+import { ArgumentNode, GraphQLResolveInfo } from 'graphql';
+
+import { AccountLoaderArgs } from '../../loaders';
+import { getAllFieldNodesByName, isInlineFragmentPresent } from './resolve-info';
+
+function findArgumentNodeByName(argumentNodes: readonly ArgumentNode[], name: string): ArgumentNode | undefined {
+    return argumentNodes.find(argumentNode => argumentNode.name.value === name);
+}
+
+function parseAccountEncodingArgument(
+    argumentNodes: readonly ArgumentNode[],
+    variableValues: {
+        [variable: string]: unknown;
+    },
+): AccountLoaderArgs['encoding'] | undefined {
+    const argumentNode = findArgumentNodeByName(argumentNodes, 'encoding');
+    if (argumentNode) {
+        if (argumentNode.value.kind === 'EnumValue') {
+            if (argumentNode.value.value === 'BASE_58') {
+                return 'base58';
+            }
+            if (argumentNode.value.value === 'BASE_64') {
+                return 'base64';
+            }
+            if (argumentNode.value.value === 'BASE_64_ZSTD') {
+                return 'base64+zstd';
+            }
+        }
+        if (argumentNode.value.kind === 'Variable') {
+            return variableValues[argumentNode.value.name.value] as AccountLoaderArgs['encoding'];
+        }
+    } else {
+        return undefined;
+    }
+}
+
+function parseAccountDataSliceArgument(
+    argumentNodes: readonly ArgumentNode[],
+    variableValues: { [variable: string]: unknown },
+): AccountLoaderArgs['dataSlice'] {
+    const argumentNode = findArgumentNodeByName(argumentNodes, 'dataSlice');
+    if (argumentNode) {
+        if (argumentNode.value.kind === 'ObjectValue') {
+            const offsetArg = argumentNode.value.fields?.find(field => field.name.value === 'offset');
+            const lengthArg = argumentNode.value.fields?.find(field => field.name.value === 'length');
+            const length = lengthArg?.value.kind === 'IntValue' ? parseInt(lengthArg.value.value) : undefined;
+            const offset = offsetArg?.value.kind === 'IntValue' ? parseInt(offsetArg.value.value) : undefined;
+            return length !== undefined && length !== null && offset !== undefined && offset !== null
+                ? { length, offset }
+                : undefined;
+        }
+        if (argumentNode.value.kind === 'Variable') {
+            return variableValues[argumentNode.value.name.value] as DataSlice;
+        }
+    } else {
+        return undefined;
+    }
+}
+
+/**
+ * Build a set of account loader args by inspecting which fields have
+ * been requested in the query (ie. `data` or inline fragments).
+ */
+export function buildAccountLoaderArgSetFromResolveInfo(
+    args: {
+        address: Address;
+        commitment?: Commitment;
+        minContextSlot?: Slot;
+    },
+    info?: GraphQLResolveInfo,
+): AccountLoaderArgs[] {
+    const argSet: AccountLoaderArgs[] = [args];
+
+    if (info?.fieldNodes[0]?.selectionSet) {
```

Comment:
Starting with the selection set, you do a bunch of spelunking into nodes and child nodes. This seems like the time to write this as an AST visitor, right?

You wrote a method called `getAllFieldNodesByName()`. The visitor version of that:

```ts
const argSet = [args];
visit(fieldNodes[0], {
  Field(node) {
    if (node.parentNode !== fieldNodes[0]) {
      
    }
    if (node.name.value !== 'data') {
      return;
    }
    const { arguments } = node;
    const encodingArg = arguments.find(({name}) => name.value === 'encoding');
    // TODO: Assert `encodingArg` exists.
    const encoding = encodingArg.value.value;
    const dataSliceArg = arguments.find(({name}) => name.value === 'dataSlice');
    // etc...
  }
  InlineFragment(node) {
    // Determine whether it should cause a `jsonParsed` fetch or not
  }
  FragmentSpread(node) {
    // Lookup the definition and continue visiting the nodes of the spread fragment
  }
});
```

Very important: here's all the stuff your visitor can do by returning different values: https://github.com/DxCx/graphql-js/blob/async/src/language/visitor.js#L77-L81

Reply:
Done.